subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyGraphBuilder.java

```diff
@@ -385,22 +389,108 @@ private void validateGraph(ResolveState resolveState) {
                     } else if (selected.hasMoreThanOneSelectedNodeUsingVariantAwareResolution()) {
                         validateMultipleNodeSelection(module, selected);
                     }
+                    if (denyDynamicSelectors) {
+                        validateDynamicSelectors(selected);
+                    }
+                    if (denyChangingModules) {
+                        validateChangingVersions(selected);
+                    }
                 }
             } else if (module.isVirtualPlatform()) {
                 attachMultipleForceOnPlatformFailureToEdges(module);
             }
         }
     }
 
+    private static boolean isDynamic(SelectorState selector) {
+        return selector.getVersionConstraint().isDynamic();
+    }
+
+    private static boolean isSimpleRangeSelector(ResolvedVersionConstraint rvc) {
+        return isSimpleSelector(rvc, VersionRangeSelector.class);
+    }
+
+    private static boolean isSimpleExactSelector(ResolvedVersionConstraint rvc) {
+        return isSimpleSelector(rvc, ExactVersionSelector.class);
+    }
+
+    private static boolean isSimpleSelector(ResolvedVersionConstraint rvc, Class<? extends VersionSelector> selectorClass) {
+        VersionSelector preferredSelector = rvc.getPreferredSelector();
+        if (preferredSelector != null && !(selectorClass.isAssignableFrom(preferredSelector.getClass()))) {
+            return false;
+        }
+        VersionSelector requireSelector = rvc.getRequiredSelector();
+        if (requireSelector != null && !(selectorClass.isAssignableFrom(requireSelector.getClass()))) {
+            return false;
+        }
+        return true;
+    }
+
+    private void validateDynamicSelectors(ComponentState selected) {
+        List<SelectorState> selectors = ImmutableList.copyOf(selected.getModule().getSelectors());
+        if (!selectors.isEmpty()) {
+            if (selectors.stream().allMatch(DependencyGraphBuilder::isDynamic)) {
+                // when all selectors are dynamic, result is undoubtedly unstable
+                markDeniedDynamicVersions(selected);
+            } else if (selectors.stream().anyMatch(DependencyGraphBuilder::isDynamic)) {
+                checkIfDynamicVersionAllowed(selected, selectors);
+            }
+        }
+    }
+
+    private void checkIfDynamicVersionAllowed(ComponentState selected, List<SelectorState> selectors) {
+        // if there's at least one dynamic selector, it's more complicated
+        // as the only stable combination is range + exact
```

Comment:
Unfortunately, that's not enough.
To be stable, the _exact version_ must be within or higher than the range.
If it is lower, then the result is dynamic.

So a build file with 
```
configurations {
    compileClasspath {
        resolutionStrategy.failOnDynamicVersions()
    }
}

dependencies {
    implementation("org.slf4j:slf4j-api:1.7.7")
    implementation("org.slf4j:slf4j-api:[1.7.25, 1.8[")
}
```
does not trigger an error but should since it resolves `org.slf4j:slf4j-api:1.7.28` today but potentially a different version next time.

Reply:
mmm, good catch... then I'm not sure how to implement this properly...